packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,240 @@
+  }
+
+  getLastKey() {
+    return this.lastKey;
+  }
+
+  getItem(key: Key) {
+    return this.keyMap.get(key);
+  }
+}
+
+let whitespace = /\s/;
+
+export function useComboBoxState<T>(props: ComboBoxProps<T>): ComboBoxState<T> {
+  let itemsControlled = !!props.onFilter;
+  let collator = useCollator({sensitivity: 'base'});
+  /*
+  let menuControlled = props.isOpen !== undefined;
+  let valueControlled = props.inputValue !== undefined;
+  let selectedControlled = !!props.selectedKey;
+   */
+
+  let selectState = useSelectState(props);
+
+  let selectedKeyItem = props.selectedKey ? selectState.collection.getItem(props.selectedKey) : undefined;
+  let selectedKeyText = selectedKeyItem ? selectedKeyItem.textValue || selectedKeyItem.rendered as string : undefined;
+  // Maybe don't need to do for defaultSelectedKey? Need to do for selectedKey so that the textfield is properly controlled reflects selectedKey text
+  let defaultSelectedKeyItem = props.defaultSelectedKey ? selectState.collection.getItem(props.defaultSelectedKey) : undefined;
+  let defaultSelectedKeyText = defaultSelectedKeyItem ? defaultSelectedKeyItem.textValue || defaultSelectedKeyItem.rendered as string : undefined;
+  // Double check if props.selectedKey should make textfield value controlled
+  let [value, setValue] = useControlledState(toString(props.inputValue) || selectedKeyText, toString(props.defaultInputValue) || defaultSelectedKeyText || '', props.onInputChange);
+  let lowercaseValue = value.toLowerCase();
+
+  let defaultFilterFn = useMemo(() => (node: Node<T>) => {
+    let scan = 0;
+    let lowercaseNode = node.textValue.toLowerCase();
+    for (let i in lowercaseValue) {
+      if (whitespace.test(lowercaseValue[i])) {
+        continue;
+      }
+      let match = false;
+      for (; scan < lowercaseNode.length && !match; scan++) {
+        if (!whitespace.test(lowercaseNode[scan])) { // if whitespace, move on to next character
+          let compareVal = collator.compare(lowercaseValue[i], lowercaseNode[scan]);
+          if (compareVal !== 0) { // if it didn't match, we can exit
+            return false;
+          } else { // we matched, move to next character in input for comparison
+            match = true;
+          }
+        }
+      }
+    }
+    return true;
+  }, [value, collator]);
+
+  selectState.collection = useMemo(() => {
+    if (itemsControlled || value === '') {
+      return selectState.collection;
+    }
+    // should value be textValue?
+    return new FilteredCollection(selectState.collection, defaultFilterFn);
+  }, [selectState.collection, value, itemsControlled, defaultFilterFn]);
+
+  useEffect(() => {
+    if (selectState.isOpen && selectState.collection.size === 0) {
+      selectState.close();
+    }
+  }, [selectState.isOpen, selectState.collection, selectState]);
+
+
+  let open = () => {
+    if (props.isFocused) {
+      selectState.open();
+    }
+  };
+
+  let toggle = (strategy, force = false) => {
```

Comment:
unsure about changing this function signature, but don't have a better solution right now